miniscope_io/uart_daq.py

```diff
@@ -1,211 +1,351 @@
+                        frame_data[i] = Bits(
+                            int=0, length=npix_expected * self.pix_depth
+                        )
+                        nbit_lost += npix_expected
+                        continue
+                    npix_header = header_data["PIXEL_COUNT"]
+                    npix_actual = len(fm_dat) / self.pix_depth
+
+                    if npix_actual != npix_expected:
+                        if i < len(self.buffer_npix) - 1:
+                            locallogs.warning(
+                                "Pixel count inconsistent for frame {} buffer {}. Expected: {}, Header: {}, Actual: {}".format(
+                                    header_data["FRAME_NUM"],
+                                    header_data["FRAME_BUFFER_COUNT"],
+                                    npix_expected,
+                                    npix_header,
+                                    npix_actual,
+                                )
+                            )
+                        nbit_expected = npix_expected * self.pix_depth
+                        if len(fm_dat) > nbit_expected:
+                            fm_dat = fm_dat[:nbit_expected]
+                        else:
+                            nbit_pad = nbit_expected - len(fm_dat)
+                            fm_dat = fm_dat + Bits(int=0, length=nbit_pad)
+                            nbit_lost += nbit_pad
+
+                    frame_data[i] = fm_dat
+
+                pixel_vector = frame_data[0]
+                for d in frame_data[1:]:
+                    pixel_vector = pixel_vector + d
+
+                assert len(pixel_vector) == (
+                    self.frame_height * self.frame_width * self.pix_depth
+                )
+
+                if self.LSB:
+                    pixel_vector = Array(
+                        "uint:32",
+                        [
+                            pixel_vector[i : i + 32][::-1].uint
+                            for i in range(0, len(pixel_vector, 32))
+                        ],
+                    )
+                img = np.frombuffer(pixel_vector.tobytes(), dtype=np.uint8)
+                imagearray.put(img)
+
+                locallogs.info(
+                    "frame: {}, bits lost: {}".format(
+                        header_data["FRAME_NUM"], nbit_lost
+                    )
+                )
 
     # COM port should probably be automatically found but not sure yet how to distinguish with other devices.
-    def capture(self, comport:str = 'COM3', baudrate:int = 1200000, mode:str = 'DEBUG'):
-        logdirectories = ['log', 'log/uart_recv', 'log/format_frame', 'log/buffer_to_frame']
+    def capture(
+        self,
+        source: str,
+        comport: str = "COM3",
+        baudrate: int = 1200000,
+        mode: str = "DEBUG",
+        bit_file: str = "/Users/mbrosch/Downloads/USBInterface/USBInterface-6mhz-3v3-INVERSE.bit",
```

Comment:
that's a last-week excitement mistake as it definitely shouldn't contains Marcel's name. It's supposed to point to a relative path within the repo so it will come with the script, but that's no longer possible. I think this can definitely be in a config file, but we just need to make sure users somehow obtain that file and put it in correct location.